### PR TITLE
proposal: Expose single annotations via downward API

### PR DESCRIPTION
**Pull request by nhlfr**
Moved from kubernetes/kubernetes#36058

Ref kubernetes/kubernetes#31218

### CODE REVIEW

**Comment by thockin** [comment]
@smarterclayton any concerns?  Yours to merge.

File: contributors/design-proposals/annotations-downward-api.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+The primary usecase for this proposal are StatefulSets. There is an idea to expose
+StatefulSet index to the applications running inside the pods managed by StatefulSet.
+Since StatefulSet creates pods as the API objects, passing this index as an
+annotation seems to be a valid way to do this. However, to finally pass this
+information to the containerized application, we need to pass this annotation.
+That's why the downward API for annotations is needed here.
+
+## API
+
+The exact `fieldPath` to the annotation will look like:
```
Comment (smarterclayton) [inline contributors/design-proposals/annotations-downward-api.md]: Describe what the syntax for fieldPath will be - JSONPath as defined in our output format.  Note that some annotations may contain the following characters `[]"'` which JSONPath has to tolerate.  We should be explicit in this proposal that we are going to use a subset of JSONPath to address field paths.

**Comment by thockin** [comment]
remind me again how an annotation KEY can every end up with [] inside it,
when
https://github.com/kubernetes/kubernetes/blob/master/pkg/api/validation/validation.go#L90

File: contributors/design-proposals/annotations-downward-api.md (hunk 1 above)
Comment (thockin) [inline contributors/design-proposals/annotations-downward-api.md]: If we all agree that the downward API keys are JSONpath, and JSONPath uses single-quotes, I am OK using single-quotes here.  It's not clear if that is a documented syntactical thing or just emergent convention.

Comment (nhlfr) [inline contributors/design-proposals/annotations-downward-api.md]: As @thockin said, annotation keys (our JSONPath) cannot contain `[]` characters. And I think that we will validate this JSONPath in the same way.

But of course I can agree with using single quotes, so the syntax will be like `metadata.annotations['annotationKey']`.

**Comment by thockin** [comment]
I don't know JSONpath well - @smarterclayton can we safely say all such
field-path strings will be JSONPath?

**Comment by bgrant0607** [comment]
cc @pmorie

File: contributors/design-proposals/annotations-downward-api.md (hunk 1 above)
Comment (smarterclayton) [inline contributors/design-proposals/annotations-downward-api.md]: JSONPath has escaping rules, we should call them out.  Not all maps are going to have limited key spaces, so better to define up front and enforce the rules rather that realize we screwed up later.

As far as I can tell JSONPath uses '' canonically.

File: contributors/design-proposals/annotations-downward-api.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+```
+metadata.annotations[annotationKey]
```
Comment (smarterclayton) [inline contributors/design-proposals/annotations-downward-api.md]: Give examples.

**Comment by smarterclayton** [comment]
A few more comments.

File: contributors/design-proposals/annotations-downward-api.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+```
+
+So, assuming that we would want to pass the `pod.beta.kubernetes.io/petset-index`
+annotation as a `PETSET_INDEX` variable, the environment variable definition
+will look like:
+
+```
+env:
+  - name: PETSET_INDEX
```
Comment (kow3ns) [inline contributors/design-proposals/annotations-downward-api.md]: Needs to be updated to reflect StatefulSet

**Comment by smarterclayton** [comment]
@liggitt we should spec this out here so we can close it out

File: contributors/design-proposals/annotations-downward-api.md (hunk 1 above)
Comment (nhlfr) [inline contributors/design-proposals/annotations-downward-api.md]: OK, I provided an explicit rule. I don't have much knowledge about JSONPath as well, so @smarterclayton please let me know whether what I wrote is correct and makes sense for you.

**Comment by nhlfr** [comment]
@smarterclayton Does it look good for you in the current form?

**Comment by smarterclayton** [comment]
Yes, this LGTM.  I think we have consensus and we haven't introduced anything net new.  When we do the impl we'll want to have a good set of sets that verify we have a consistent and concise form.

**Comment by deitch** [comment]
Ah, this makes me very happy. Having to extend things like zookeeper just to work around this was wasteful. 

Is there a k8s target version and release date?

**Comment by mikesparr** [comment]
Hi, any status on this? Will this PR solve the ordinal index issue? #40651

@nhlfr and @smarterclayton  were working on the petset issue #30427 and then I assume that got absorbed into stateful set issues.

I'm on `1.8.4` gke and still cannot get access to the ordinal index of a Stateful Set for dynamic `env` params for zookeeper, kafka, etc. clusters.  Instead of all the shell and image hacks people are doing to solve this, if we simply passed the ordinal index and made it accessible in `env` of config, it would clean up deployments.

**Comment by mikesparr** [comment]
Any update on this and will it solve https://github.com/kubernetes/kubernetes/issues/30427 and https://github.com/kubernetes/kubernetes/issues/40651

Originally was earmarked for `1.5` and I'm on `1.8.4` and still no luck. Trying various ways to get ordinal index for ENV for container with no luck (aside from having to rebuild containers and start manually using command and `${HOSTNAME##*-}` hack).

Goal:
screenshot 2017-12-11 09 40 57

**Comment by mikesparr** [comment]
I was able to solve until the downstream feature available using lifecycle hooks and during postStart, exec a command to export the env param to container. Just in case anyone else running into this, here was solution:

```
        lifecycle:
          postStart:
            exec:
              command: ["/bin/sh", "-c", "export KAFKA_BROKER_ID=${HOSTNAME##*-}"]
```

**Comment by yellowmegaman** [comment]
Any news on this? Seems there's still no way to do it properly, except hack with hostname.

**Comment by coderanger** [comment]
@yellowmegaman This proposal was implemented long ago, it works as stated in the document.

**Comment by yellowmegaman** [comment]
@coderanger thanks!
Sorry, somehow github interface tricked me into thinking that there were additions after merge.
Awesome!)

**Comment by yellowmegaman** [comment]
```
        env:
          - name: STATEFULSET_INDEX
            valueFrom:
              fieldRef:
                fieldPath: metadata.annotations['spec.pod.beta.kubernetes.io/statefulset-index']
```

Didn't work, gonna dig it later then

**Comment by mikesparr** [comment]
Agree, this has never worked for me either, so I resorted to using
lifecycle hooks or custom shell scripting to create the necessary env vars
for clustered pods. It would be great to not have to use workarounds. I was
testing before on 1.9.6. I am now on 1.12.5 and haven't tested but
previously they did not work.

**Comment by thockin** [comment]
https://github.com/kubernetes/kubernetes/pull/55902 Was supposed to
implement this.  There's even tests -- for the validation.

I just tried to use it and, yes, it seems not to work.  I'm opening a new
issue: https://github.com/kubernetes/kubernetes/issues/75555

Please take conversation there

**Comment by thockin** [comment]
I was wrong - it does work, it seems.   I screwed up the test.

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: test-downapi-annotations
spec:
  selector:
    matchLabels:
      run: test-downapi-annotations
  template:
    metadata:
      labels:
        run: test-downapi-annotations
      annotations:
        testann: "value of testann"                                                                                                                                                     
    spec:
      volumes:
      - name: annvol
        downwardAPI:
          items:
          - path: testann
            fieldRef:
              fieldPath: metadata.annotations['testann']
      containers:
      - name: app
        image: k8s.gcr.io/serve_hostname:v1.4
        env:
        - name: TESTANN
          valueFrom:
            fieldRef:
              fieldPath: metadata.annotations['testann']
        volumeMounts:
        - mountPath: /downward-api
          name: annvol
```

**Comment by yellowmegaman** [comment]
@thockin thatnks for shedding the light! Do you have, by any chance, understanding how to make stateful-index work too? cause it ain't working. 
Very active googling returns only examples just I've submitted earlier

**Comment by yellowmegaman** [comment]
Even tests suggest that it is correct naming.
https://github.com/kubernetes/kubernetes/pull/55902/commits/82c9eec1648aa8ce342be6f3c799dcc62dd47070
```
        env:
          - name: STATEFULSET_INDEX
            valueFrom:
              fieldRef:
                fieldPath: metadata.annotations['spec.pod.beta.kubernetes.io/statefulset-index']
```
```
root@cassandra-0:/# env | grep -i index
STATEFULSET_INDEX=
root@cassandra-0:/#
```

**Comment by liggitt** [comment]
I don't see anything setting that annotation on stateful set pods... can you get a statefulset pod from the API and see what is actually in its metadata?

**Comment by yellowmegaman** [comment]
Sorry if I misunderstood what exactly you want.

## pod

```
$ kubectl -n db describe po cassandra-0
Name:               cassandra-0
Namespace:          db
Priority:           0
PriorityClassName:  <none>
Node:               gke-propa-scalenodes-fe046b0e-wlhm/10.164.0.33
Start Time:         Fri, 22 Mar 2019 17:17:11 +0300
Labels:             app=cassandra
                    controller-revision-hash=cassandra-55c4656855
                    statefulset.kubernetes.io/pod-name=cassandra-0
Annotations:        traffic.sidecar.istio.io/excludeOutboundIPRanges: 0.0.0.0/0
Status:             Running
IP:                 10.60.2.45
Controlled By:      StatefulSet/cassandra
Containers:
  cassandra:
    Container ID:   docker://0a8363bf9bb253017feb178f0e11a43e1d808bf6afe97607dfef5bfcf3cc0357
    Image:          quay.io/yellowmegaman/container-cassandra:buster-3.11.4
    Image ID:       docker-pullable://quay.io/yellowmegaman/container-cassandra@sha256:8d3ea96041dc991d73bf4562428c406b3eb11a03bfb103fa46ed5409f0be2c97
    Ports:          7000/TCP, 7001/TCP, 7199/TCP, 9042/TCP
    Host Ports:     0/TCP, 0/TCP, 0/TCP, 0/TCP
    State:          Running
      Started:      Fri, 22 Mar 2019 17:17:22 +0300
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     500m
      memory:  3Gi
    Requests:
      cpu:      500m
      memory:   3Gi
    Readiness:  exec [nodetool status] delay=30s timeout=5s period=10s #success=1 #failure=3
    Environment:
      STATEFULSET_INDEX:            (v1:metadata.annotations['spec.pod.beta.kubernetes.io/statefulset-index'])
      MAX_HEAP_SIZE:               2048M
      HEAP_NEWSIZE:                100M
      CONFIG_TEMPLATE_LOCATION:    /opt/cassandra/cassandra.yaml
      CASSANDRA_CLUSTER_NAME:      k8s
      CASSANDRA_DATA_DIR:          /var/lib/cassandra/data
      CASSANDRA_COMMITLOG_DIR:     /var/lib/cassandra/commitlog
      CASSANDRA_SAVED_CACHES_DIR:  /var/lib/cassandra/saved_caches
      CASSANDRA_SEED_NODE:         cassandra-0.cassandra.db.svc.cluster.local
      CASSANDRA_LISTEN_ADDRESS:     (v1:status.podIP)
      CASSANDRA_START_RPC:         true
      CASSANDRA_RPC_ADDRESS:        (v1:status.podIP)
      CHMOD:                       true
      CHMODDIR:                    /var/lib/cassandra
    Mounts:
      /etc/podinfo from podinfo (rw)
      /opt/cassandra from cassandra-config (rw)
      /var/lib/cassandra from cassandra-data (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-6sm99 (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             True 
  ContainersReady   True 
  PodScheduled      True 
Volumes:
  cassandra-data:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  cassandra-data-cassandra-0
    ReadOnly:   false
  cassandra-config:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      cassandra-config
    Optional:  false
  podinfo:
    Type:  DownwardAPI (a volume populated by information about the pod)
    Items:
      metadata.labels -> labels
      metadata.annotations -> annotations
  default-token-6sm99:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-6sm99
    Optional:    false
QoS Class:       Guaranteed
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type    Reason     Age   From                                         Message
  ----    ------     ----  ----                                         -------
  Normal  Scheduled  102s  default-scheduler                            Successfully assigned db/cassandra-0 to gke-propa-scalenodes-fe046b0e-wlhm
  Normal  Pulling    93s   kubelet, gke-propa-scalenodes-fe046b0e-wlhm  pulling image "quay.io/yellowmegaman/container-cassandra:buster-3.11.4"
  Normal  Pulled     92s   kubelet, gke-propa-scalenodes-fe046b0e-wlhm  Successfully pulled image "quay.io/yellowmegaman/container-cassandra:buster-3.11.4"
  Normal  Created    91s   kubelet, gke-propa-scalenodes-fe046b0e-wlhm  Created container
  Normal  Started    91s   kubelet, gke-propa-scalenodes-fe046b0e-wlhm  Started container
```
## statefuleset
```
$ kubectl -n db describe sts cassandra
Name:               cassandra
Namespace:          db
CreationTimestamp:  Fri, 22 Mar 2019 17:17:00 +0300
Selector:           app=cassandra
Labels:             app=cassandra
Annotations:        kubectl.kubernetes.io/last-applied-configuration:
                      {"apiVersion":"apps/v1","kind":"StatefulSet","metadata":{"annotations":{},"labels":{"app":"cassandra"},"name":"cassandra","namespace":"db"...
Replicas:           824639463496 desired | 3 total
Update Strategy:    RollingUpdate
  Partition:        824639464476
Pods Status:        2 Running / 1 Waiting / 0 Succeeded / 0 Failed
Pod Template:
  Labels:       app=cassandra
  Annotations:  traffic.sidecar.istio.io/excludeOutboundIPRanges: 0.0.0.0/0
  Containers:
   cassandra:
    Image:       quay.io/yellowmegaman/container-cassandra:buster-3.11.4
    Ports:       7000/TCP, 7001/TCP, 7199/TCP, 9042/TCP
    Host Ports:  0/TCP, 0/TCP, 0/TCP, 0/TCP
    Limits:
      cpu:     500m
      memory:  3Gi
    Requests:
      cpu:      500m
      memory:   3Gi
    Readiness:  exec [nodetool status] delay=30s timeout=5s period=10s #success=1 #failure=3
    Environment:
      STATEFULSET_INDEX:            (v1:metadata.annotations['spec.pod.beta.kubernetes.io/statefulset-index'])
      MAX_HEAP_SIZE:               2048M
      HEAP_NEWSIZE:                100M
      CONFIG_TEMPLATE_LOCATION:    /opt/cassandra/cassandra.yaml
      CASSANDRA_CLUSTER_NAME:      k8s
      CASSANDRA_DATA_DIR:          /var/lib/cassandra/data
      CASSANDRA_COMMITLOG_DIR:     /var/lib/cassandra/commitlog
      CASSANDRA_SAVED_CACHES_DIR:  /var/lib/cassandra/saved_caches
      CASSANDRA_SEED_NODE:         cassandra-0.cassandra.db.svc.cluster.local
      CASSANDRA_LISTEN_ADDRESS:     (v1:status.podIP)
      CASSANDRA_START_RPC:         true
      CASSANDRA_RPC_ADDRESS:        (v1:status.podIP)
      CHMOD:                       true
      CHMODDIR:                    /var/lib/cassandra
    Mounts:
      /etc/podinfo from podinfo (rw)
      /opt/cassandra from cassandra-config (rw)
      /var/lib/cassandra from cassandra-data (rw)
  Volumes:
   cassandra-config:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      cassandra-config
    Optional:  false
   podinfo:
    Type:  DownwardAPI (a volume populated by information about the pod)
    Items:
      metadata.labels -> labels
      metadata.annotations -> annotations
Volume Claims:
  Name:          cassandra-data
  StorageClass:  fast
  Labels:        <none>
  Annotations:   <none>
  Capacity:      9Gi
  Access Modes:  [ReadWriteOnce]
Events:
  Type     Reason            Age                   From                    Message
  ----     ------            ----                  ----                    -------
  Warning  FailedCreate      2m4s (x11 over 2m9s)  statefulset-controller  create Pod cassandra-0 in StatefulSet cassandra failed error: The POST operation against Pod could not be completed at this time, please try again.
  Normal   SuccessfulCreate  118s                  statefulset-controller  create Pod cassandra-0 in StatefulSet cassandra successful
  Normal   SuccessfulCreate  62s                   statefulset-controller  create Pod cassandra-1 in StatefulSet cassandra successful
  Normal   SuccessfulCreate  9s                    statefulset-controller  create Pod cassandra-2 in StatefulSet cassandra successful
```

Also I've used this https://kubernetes.io/docs/tasks/inject-data-application/downward-api-volume-expose-pod-information/ to add `podinfo`

```
root@cassandra-0:/# cat /etc/podinfo/annotations && echo -e '\n'
kubernetes.io/config.seen="2019-03-22T14:17:11.139177263Z"
kubernetes.io/config.source="api"
traffic.sidecar.istio.io/excludeOutboundIPRanges="0.0.0.0/0"

root@cassandra-0:/#
```
Istio is actually disabled.

**Comment by liggitt** [comment]
as I suspected, the pod does not have the annotation you are attempting to inject via the downward API:

```
Labels:             app=cassandra
                    controller-revision-hash=cassandra-55c4656855
                    statefulset.kubernetes.io/pod-name=cassandra-0
Annotations:        traffic.sidecar.istio.io/excludeOutboundIPRanges: 0.0.0.0/0
```

that's why no value is injected

**Comment by yellowmegaman** [comment]
@liggitt thank you, it helped a lot, but after lots of time spent searching, still can't find a way to set this annotation properly.

Ended up with:
```
---
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: cassandra
  namespace: db
  labels:
    app: cassandra
spec:
  serviceName: cassandra
  replicas: 3
  selector:
    matchLabels:
      app: cassandra
  template:
    metadata:
      annotations:
        traffic.sidecar.istio.io/excludeOutboundIPRanges: 0.0.0.0/0
        spec.pod.beta.kubernetes.io/statefulset-index: "true"
      labels:
        app: cassandra
...
```
Which gives:
```
root@cassandra-0:/# env | grep -i index
STATEFULSET_INDEX=true
root@cassandra-0:/#
```

But how can i set it properly to get actual index?
Sorry for bothering, but there's no hint in tests / threads here on GH. (or I'm blind).

**Comment by liggitt** [comment]
> @liggitt thank you, it helped a lot, but after lots of time spent searching, still can't find a way to set this annotation properly.

statefulset does not expose this information currently. there are tracking issues (https://github.com/kubernetes/kubernetes/issues/30427 and https://github.com/kubernetes/kubernetes/issues/40651) and a [WIP PR](https://github.com/kubernetes/kubernetes/pull/68719) to add it.

once statefulset actually populates the annotation, you can use the downward API to inject it into your pod

**Comment by yellowmegaman** [comment]
@liggitt Thanks for your patience!

**Comment by smarterclayton** [comment]
I've seen a fair amount of interest in this recently from a number of users, and some use cases have also made labels necessary.  @ahg opened https://github.com/kubernetes/kubernetes/issues/117986, I think a quick KEP is in order, and I think folks here who have use cases would be exactly the input we need to get the basic set locked down.  Please do not hesitate to suggest in that issue, and we'll look through this one for more inputs.

**Pull request merged**